### ISSUE TITLE
Serialization Cache for Classes does not work

### ISSUE THREAD

**Issue by mkaesemann** [larshp/abapGit#1021]
I'm trying to use abapGit for a rather large add-on development (>11000 objects) and reading and serialization takes 5-10 minutes each time I switch to the repo, so I'm investigating how to improve performance there.
During the initial debugging I came across the serialization caching mechanism that is implemented in `lcl_repo->get_files_local` and noticed that the caching for classes always fails.
Upon analysis I found that this was caused because the `lcl_object_class-lif_object~has_changed_since` determines different includes for the class than actually exist in the system and thus the `check_prog_changed_since` method always returns true.
I can see in `mo_object_oriented_object_fct->get_includes` that some includes are returned independently of whether they exist in the system or not:
```abap
    APPEND cl_oo_classname_service=>get_ccdef_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_ccmac_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_ccimp_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_cl_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_ccau_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_pubsec_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_prosec_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_prisec_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_classpool_name( lv_class_name ) TO rt_includes.
    APPEND cl_oo_classname_service=>get_ct_name( lv_class_name ) TO rt_includes.
```
and was wondering what the reasoning behind this determination is.
Why are they assumed to be there instead of aligned against what is actually available in the system (e.g. with a check against `REPOSRC`)?
Is this done on purpose or a bug?
I assumed it could have something to do with being unable to detect deletions, but I'm not sure if that actually factors in or not.
Can you provide some insight so I know if it is worth analysing this further?

**Comment by larshp** [comment]
It can be regarded as a but, however I think it is slower to figure out if the include exists or not.

method `lif_object~changed_by` also calls the `get_includes` method, which does FOR ALL ENTRIES instead of single selection.

I think the easiest and most performant fix to this is rewriting `has_changed_since` to use FOR ALL ENTRIES instead. Determining if the include exists or not would anyhow have to lookup in REPOSRC

**Comment by mkaesemann** [comment]
I see what you mean and it is probably worth measuring before actually doing any changes. Especially since one has to factor in the time the serialization of the class takes (a longer running detection might be compensated by a skipped serialization).

In my fork I have already started working on retaining the buffer of serialized data between session (in the DB), which also improves performance for large projects, so I basically already have a comparison state I can use to detect if something has been serialized before, so it's probably easier to do that in the official version.

Maybe I'll check the FOR ALL ENTRIES idea seperately in another fork - but probably not prior to the TechEd due to work.

**Comment by larshp** [comment]
@sbcgua any thoughts on removing this feature? HAS_CHANGED_SINCE
It will make it easier to implement parallel serialization(see #1714)

**Comment by sbcgua** [comment]
@larshp
Didn't look deeply so the below may be a bit superficial

- when we created the cache the serialization performance improved. So in general I'd be for leaving the feature (although now it is broken indeed, i noticed the speed downgrade too :((( )
- Clearly caching is a kind optimization which make certain changes a lot more complex. The code changed so the caching should probably change too. Ideally the caching should be a transparent layer which can be easily disabled e.g. by specific customer settings.
- `has_changed_since` thus will remain (as the logic which identifies the fact of modification), but the way it is used will change
- I'd also think about some standartized performance test right in the abapGit code (kind of unit test, but for the benchmarking, so that results are comparable, also can useful in general to collect stats about situation in different envs). Maybe worth to open a separate issue for this.

Sorry for so superficial reply ... I'll try to look a bit deeper this or next week if needed and come with more concrete ideas.

**Comment by joymike** [comment]
I have tested the suggestion above, and toyed around with my fork with extending the user session caching into DB. (i.e. persisting ZCL_ABAPGIT_REPO-mt_local and ZCL_ABAPGIT_REPO-mv_last_serialization). When the repo is instantiated next, I read from DB and populate mt_local and mv_last_serialization.
I reduced the load time of 4-5 mins down to 2 mins. (8000 objects)
There seems to be an existing feature to force local refresh which invalidates the cache if incorrect.

I must be missing something. I don't think there's anything wrong with the concept of caching or has_changed_since. I see more of a problem of the implementation of has_changed_since by individual objects. 

Upon a bit of debugging, I noticed:
- As @mkaesemann pointed out, program assumes includes of classes that doesn't exist. ZCL_ABAPGIT_OBJECTS_PROGRAM=>check_prog_changed_since does not check for sy-subrc after the select on REPOSRC. Hence the class is always considered as changed if it doesnt have all the includes.
- similar problem in "TABL" when checking for technical settings dd09l for ddic structures which is also considered TABL - it does not check sy-subrc again. Of course it will not find it and consider it changed.
- similar problem in "VIEW" when checking for technical settings dd09l for views. I suspect this is a copy and paste error. Of course it will not find it and consider it changed.

I will try to fix these this week (and create my first PR).

By fixing these, I managed to get the load time to down to approx a minute now. There are still approx 1000 objects that does not have a timestamp concept or perhaps has_changed_since is not implemented correctly - meaning it has to always serialise them.

**Comment by larshp** [comment]
If the cache is persisted to db, then how will changes to serializing code be handled? I.e. a field is removed or added in the serializer

**Comment by sbcgua** [comment]
Imho. Cache in DB is a bit an overkill. It might bring more issues than solve.

Open transports: what if the development is local ?

BTW what about ADT? Does the session persist while client side is running (I just don't know, sorry :) ? Or is it restarted on each call? This may add additional strong pro or cons to the caching.

**Comment by larshp** [comment]
open transports: yeah, it would be a setting on the repo, I'm assuming that a lot of these very large repos are in-house developments

ADT: is stateless as far as I know

**Comment by joymike** [comment]
I'm hearing reservations with the idea. Although with obvious performance improvement as I have proved in my system, there's always an element of risk that the cache being out of date with the actual system causing problems.

@larshp , your scenario is valid but unlucky. Maybe I'm wrong, won't the scenario you are describing already exist with the cache implementation in the repository class today? 

By comparing the object's "changed timestamp" with the last serialized time, provided they resync/restart the transaction it should show the change.  This of course assumes the has_changed_since implementation is correct. My initial thought was to make this an optional repository parameter, so that the user can choose whether to use it or not. Default to be off (ie no caching in DB). 
 
However, I agree this approach might be error prone compared to parallel serialization. 
I'll hold off making changes to storing in DB until I see the parallel serialization feature. Parallel serialization feature with function group is not compatible with the copy and paste install approach I guess where as caching is. 

Thanks for your comments above.. Really enjoyed such constructive discussion.
In the meantime I might go and fix a things in the has_changed_since code I pointed out.

**Comment by larshp** [comment]
cache implementation today: yes, but when restarting abapGit the problem is gone, as cache is not persisted

**Comment by flaiker** [comment]
Regarding parallelization and copy-paste-compatibility:
On newer releases there is CL_ABAP_PARALLEL, which allows you to do stuff in parallel, without having to create function modules or customize parallel processing frameworks (-> it should be copy-paste-compatible). Maybe the internals are already there on 7.02 so one could backport it.

**Comment by larshp** [comment]
CL_ABAP_PARALLEL: thanks for the hint, I'll look into that one

regarding cache: I think there is a bug regarding the inactive markings, if the object is in cache, and the object is inactive, it does not show that it is inactive

**Comment by sbcgua** [comment]
> ADT: is stateless as far as I know

This factor is against the in-memory caching then :((

**Comment by larshp** [comment]
again I'll suggest removing this feature, as established above there are multiple issues with it which makes it look like a problem in abapGit, ie. on my local system the CLAS objects are refreshed on refresh(due to above bug) but INTFs are not refreshed(due to bad local settings)

Also now parallel serialization is implemented, which gives a performance increase.

**Comment by larshp** [comment]
Also guess if objects are translated this is not reflected when refreshing

**Comment by sbcgua** [comment]
Looks like factors against caching accumulate :)
Since parallel serialization my biggest package serializes in reasonable time (really cool feature). so from purely selfish point of view I personally won't suffer a lot from cache removing for now.
P.S. BTW I think remote fetch can also go to parallel task, this would make UI subjectively faster. But I don't know how to ask for the password from the thread - do you know if it is possible ?

**Comment by larshp** [comment]
parallel password: guess it will be difficult

the best/next performance improvements would be saving git locally #357, plus only serializing objects part of open/recent transports(as a repo setting?)

**Comment by mkaesemann** [comment]
I was already looking serializing only objects of open/recent transports, but quickly discarded the idea, because it doesn't cover depending objects. For example, if you change a data element, only the data element is registered in the transport, but technically all structures, tables, etc. that use the data element are also affected and change.
So unless you want to implement a check which objects, that are not in the transport, are affected by a change of an object in the transport, I don't think that option leads to a consistent repo.

**Comment by larshp** [comment]
@mkaesemann yeah, true, good observation. We should try working towards file formats with non-cascading changes, however I dont know if it is feasible at all.

**Comment by larshp** [comment]
merged, closing
